// @inkjs/ui@1.0.0/build/table/theme.js
const theme = {
    parts: [
        'table',
        'header',
        'headerCell',
        'headerValue',
        'cell',
        'value',
        'body',
        'row',
    ],
    baseStyle: {
        table: {
            flexDirection: 'column',
        },
        header: {
            flexShrink: 0,
            borderStyle: 'single',
            borderTop: false,
            borderLeft: false,
            borderRight: false,
            gap: 1,
        },
        headerCell: {
            flexShrink: 1,
            flexGrow: 1,
            flexBasis: '0%',
            minWidth: 0,
        },
        headerValue: {
            wrap: 'truncate-end',
        },
        row: {
            gap: 1,
        },
        cell: {
            flexShrink: 1,
            flexGrow: 1,
            flexBasis: '0%',
            minWidth: 0,
        },
        value: ({ isFocused }) => ({
            underline: isFocused,
            wrap: 'truncate-end',
        }),
        body: {
            flexGrow: 1,
            minHeight: 0,
            flexDirection: 'column',
        },
    },
};
export default theme;
