// @inkjs/ui@1.0.0/build/status-message/theme.js
import figures from 'figures';
const theme = {
    styles: {
        container: () => ({
            gap: 1,
        }),
        iconContainer: () => ({
            flexShrink: 0,
        }),
        icon({ variant }) {
            let color;
            if (variant === 'info') {
                color = 'blue';
            }
            if (variant === 'success') {
                color = 'green';
            }
            if (variant === 'error') {
                color = 'red';
            }
            if (variant === 'warning') {
                color = 'yellow';
            }
            return {
                color,
            };
        },
        message: () => ({}),
    },
    config({ variant }) {
        let icon;
        if (variant === 'info') {
            icon = figures.info;
        }
        if (variant === 'success') {
            icon = figures.tick;
        }
        if (variant === 'error') {
            icon = figures.cross;
        }
        if (variant === 'warning') {
            icon = figures.warning;
        }
        return { icon };
    },
};
export default theme;
